v4.0.0-dev — API refactor, caching, fallback fixes, and benchmark suite

aissat opened the pull request:
# v4.0.0-dev — API refactor, caching, fallback fixes, and benchmark suite

## Summary

This PR modernizes the `easy_localization` API by moving `path` and `supportedLocales` into `AssetLoader`, adding built-in caching, introducing a clean storage interface with `SharedPreferencesStorage` and `InMemoryStorage`, fixing fallback translation bugs, and introducing a comprehensive benchmark suite (v3 vs v4).

---

## ⚠️ Breaking Changes

### `AssetLoader` rework

`AssetLoader` is no longer a stateless pass-through. It now owns `path` and `supportedLocales`, passed via constructor:

```dart
// v3 — stateless, path on widget
abstract class AssetLoader {
  Future<Map<String, dynamic>> load(String localePath);
}

// v4 — owns path + locales
abstract class AssetLoader {
  final String? path;
  final List<Locale>? supportedLocales;
  const AssetLoader({this.path, this.supportedLocales});

  Future<Map<String, dynamic>> load({Locale? locale});
}
```

### `RootBundleAssetLoader` requires `path`

```dart
// v3
const RootBundleAssetLoader()

// v4
const RootBundleAssetLoader(path: 'assets/translations')
```

### `EasyLocalization` widget simplified

`s`supportedLocales`` and `path` are no longer widget properties — they live on the loader. The widget only receives `child`, `fallbackLocale`, `startLocale`, `saveLocale`, `useFallbackTranslations`, `useOnlyLangCode`, and `errorWidget`.

### `ensureInitialized()` is now required, takes loader + optional storage

```dart
await EasyLocalization.ensureInitialized(
  assetLoader: const RootBundleAssetLoader(path: 'assets/translations'),
  storage: const SharedPreferencesStorage(), // optional, defaults to SharedPreferencesStorage
);
```

### Removed APIs

| API | Reason |
|---|---|
| `useFallbackTranslationsForEmptyResources` | Unclear semantics, rarely used |
| `ignorePluralRules` | Relied on removed ICU data |
| `extraAssetLoaders` | Complex chaining, better solved by custom `AssetLoader` |
| `AssetLoader.localeExists()` | `load()` throws on failure instead |

---

## 🗄️ Storage Interface

Provided a clean, testable storage abstraction with two implementations:

### `IEasyLocalizationStorage`
```dart
abstract class IEasyLocalizationStorage {
  Future<void> init();
  Future<String?> getValue(String key);
  Future<void> setValue(String key, String value);
  Future<void> removeValue(String key);
  Future<void> close();
}
```

### `SharedPreferencesStorage` (default)

Backed by `shared_preferences`. Pass to `ensureInitialized()`:

```dart
await EasyLocalization.ensureInitialized(
  assetLoader: const RootBundleAssetLoader(path: 'assets/translations'),
  storage: const SharedPreferencesStorage(),
);
```

### `InMemoryStorage` (testing)

Ephemeral map-based storage for unit tests — no mock setup needed:

```dart
final storage = InMemoryStorage();
await EasyLocalization.ensureInitialized(
  assetLoader: const RootBundleAssetLoader(path: 'assets/translations'),
  storage: storage,
);

print(await storage.getValue('locale')); // null
await storage.setValue('locale', 'en_US');
print(await storage.getValue('locale')); // 'en_US'
```

---

## ✨ New Features

### Built-in caching

Three caching layers available:

| Class | Description |
|---|---|
| **`CachedAssetLoader`** | Base class with static `_translationCache`, `cacheTranslations()`, `getCachedTranslations()`, `isCached()`. Building block for custom loaders. |
| **`OptimizedAssetLoader`** | Extends `RootBundleAssetLoader` with a separate static cache layer + cache-hit logging. |
| **`RootBundleAssetLoader`** | Now caches translations per locale in a static `_cache` internally. Switching back to a previously loaded locale costs ~0.02–0.09 ms. |

### Interactive switch timing

When running `flutter run` and tapping a language:

- **SnackBar**: `Switch time: 0.56 ms`
- **On-screen**: `Last switch: X.XX ms` on the main page
- **Logs**: `BENCH	switch (interactive: en_US)	0.56 ms`

### Benchmark suite

`flutter_benchmark/` contains identical Flutter Linux apps for v3.0.8 vs v4.0.0-dev.

Run with: `cd flutter_benchmark && bash bench.sh`

Sample results (Linux x64 release AOT):

| Benchmark | v3 (ms) | v4 (ms) | Winner |
|---|---|---|---|
| Startup → first frame | 982.41 | 976.22 | ≈ tie |
| Cold `setLocale()` | 3.69 | 3.76 | ≈ tie |
| Switch avg (10 rounds) | 0.63 | 0.06 | ▲ v4 **10× faster** |

---

## 🐛 Bug Fixes

### Fallback translations overwritten by base language

When `_locale` has a country code (e.g. `es_AR`) and `supportedLocales` contains the base language (`es`), the base language data was merged **after** the fallback locale data, overwriting real fallback values with `null`:

```dart
// BAD (v3): base data clobbered fallback
fallbackData.addAll(baseData);

// FIX (v4): skip nulls
baseData.forEach((k, v) {
  if (v != null) fallbackData.putIfAbsent(k, () => v);
});
```

### `useOnlyLangCode` with country-code locales

`_locale` was **static** — all controller instances shared it, and later constructor calls overwrote it before `loadTranslations()` ran. Restructured tests to create + load each controller within the same microtask.

### `CachedAssetLoader` missing `const` constructor

Added `const CachedAssetLoader({required super.path})` for use in const contexts.

### `ListTile` ink-splash warnings

Wrapped `Container` + `ListTile` in `Material(type: MaterialType.transparency)` across all benchmarks and example.

---

## 🧪 Test Improvements

- Rewrote `test/easy_localization_test.dart` (~175 → ~600 lines)
- New test groups: `Asset loader`, `wrong path`, linked locale messages, nested arguments, fallback edge cases
- All tests use explicit `EasyLocalizationController` instances (no global singleton)
- Fallback tests cover: base-language merge, missing fallback keys, `putIfAbsent` null-skip

---

## Migration Guide

### Step 1: Update `main()`

```dart
// Before (v3)
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(EasyLocalization(
    child: MyApp(),
    supportedLocales: [Locale('en')],
    path: 'assets/translations',
  ));
}

// After (v4)
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await EasyLocalization.ensureInitialized(
    assetLoader: const RootBundleAssetLoader(path: 'assets/translations'),
  );
  runApp(EasyLocalization(child: MyApp()));
}
```

### Step 2: Update custom asset loaders

```dart
// Before (v3)
class MyLoader extends AssetLoader {
  @override
  Future<Map<String, dynamic>> load(String path, Locale locale) async {
    return json.decode(
      await File('$path/${locale.languageCode}.json').readAsString(),
    );
  }
}

// After (v4)
class MyLoader extends AssetLoader {
  const MyLoader({required String path}) : super(path: path);

  @override
  Future<Map<String, dynamic>> load({Locale? locale}) async {
    return json.decode(
      await File('$path/${locale!.languageCode}.json').readAsString(),
    );
  }
}
```



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Pluggable asset loaders with built-in caching options and a new persistent/in-memory storage setup.
  - Migration helpers for older configurations, plus new Arabic locale assets.
  - Added Playwright CI workflow for end-to-end testing.

- **Breaking Changes**
  - Updated initialization flow to require an asset loader/storage configuration; loader API and supported locales handling changed, alongside major package/SDK constraint updates.

- **Improvements**
  - Enhanced pluralization fallback behavior and more informative translation-load errors.
  - Theme-related widget tweaks and updated default error text.

- **Bug Fixes**
  - Improved translation existence checks across fallback data.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aissat committed:
feat(v4): update example app

aissat committed:
refactoring `AssetLoader` and  `RootBundleAssetLoader`

aissat committed:
feat: create easy_localization_storage_interface.dart

aissat committed:
update easy_localization_app.dart

aissat committed:
add export storage interface

aissat committed:
update `AssetLoader` docs

aissat committed:
refactoring `EasyLocalizationController` update `initEasyLocation` supported locales data and setting the device and saved locale.

aissat committed:
update generate.dart

aissat committed:
chore: update pubspec.yaml

aissat committed:
chore: create ar.json, ar.json and en.json

aissat committed:
feat(test) :update test

MatrixRave commented:
Seems like a valid improvement to me. Would be nice to see this getting approved or any recognition.

aissat committed:
Add caching support to RootBundleAssetLoader

Introduced CachedAssetLoader for base functionality and updated
RootBundleAssetLoader to include internal translation caching. Added
OptimizedAssetLoader as a wrapper for backward compatibility.

aissat committed:
Refactor localization controller and improve translation loading

- Optimize translation lookup and caching
- Add `TranslationLoadException` for better error handling
- Update `TextTranslateExtension` to use `TextScaler`
- Refactor `EasyLocalizationController` for cleaner initialization
- Improve plural rule evaluation fallback

aissat committed:
Update dependencies and SDK syntax

- Upgrade intl to version 0.19.0+
- Add benchmark_harness to dev_dependencies
- Standardize SDK version quote style

aissat committed:
Add performance benchmarking and feedback to locale switch

Included a stopwatch to measure the locale switching time and displayed
the
result via a SnackBar to provide immediate performance feedback.

aissat committed:
Improve translation loading and cache safety

- Add constructor to CachedAssetLoader
- Prevent overwriting existing keys when merging fallback translations
- Log TranslationLoadException details
- Expose raw data in Translations for testing purposes

aissat committed:
Refactor asset loader tests and improve controller initialization

Updated test suite to use named arguments for RootBundleAssetLoader and
initialized controllers within setUpAll to ensure correct state.

aissat committed:
Make storage interface methods asynchronous

Update IEasyLocalizationStorage to use Future for all operations to
support async-first persistence layers like SharedPreferences. Added
concrete SharedPreferences and InMemory implementations.

aissat committed:
Update documentation and API for pluggable storage

- Refactored `EasyLocalization.ensureInitialized` to accept
  `assetLoader`
  and optional `storage`.
- Added `IEasyLocalizationStorage` interface with
  `SharedPreferencesStorage`
  and `InMemoryStorage` implementations.
- Updated README examples and widget properties to reflect the new
  initialization flow.

aissat committed:
Add compatibility extension for v3 migration

aissat committed:
Add null check to asset loader and add compatibility tests

aissat committed:
Update widget tests for localization configuration

Refactor test setup to include explicit supported locales and start
locales in EasyLocalization initializers to ensure consistent test
environments.

aissat committed:
Update changelog for 4.0.0-dev release

aissat committed:
Fix formatting and clean up metadata in CHANGELOG and pubspec

aissat committed:
Add .pubignore file to exclude unnecessary files

aissat committed:
fix: untrack .iml .metadata, unignore project.pbxproj from gitignore

aissat committed:
Remove ephemeral Flutter iOS configuration files

aissat committed:
Remove localization-kit from topics in pubspec.yaml

aissat committed:
Refactor asset loading and storage initialization

- Update CachedAssetLoader to inherit from RootBundleAssetLoader.
- Simplify OptimizedAssetLoader by utilizing base class caching.
- Add error handling for uninitialized SharedPreferences.
- Fix locale saving condition in EasyLocalizationController.

aissat committed:
Fix asset loader tests

Reset EasyLocalization state after fallback locale tests and correct the

cache clearing reference in OptimizedAssetLoader tests.

aissat committed:
Update SDK constraint to support older versions

aissat committed:
Update device locale detection to use PlatformDispatcher

aissat committed:
Update version to 4.0.0-dev and document breaking changes

aissat committed:
Add library documentation and refine exports

- Added a library-level description to easy_localization.dart.
- Restricted intl package exports to specific classes.
- Added a macro reference to the plural extension method.

aissat committed:
Add pure resolver functions for translation data

Introduces a suite of functions for resolving nested translation keys,
handling link references, and performing argument substitution. These
functions operate on plain maps and strings, ensuring they are
environment-agnostic and easy to test.

aissat committed:
Add unit tests for resolver functions

aissat committed:
Clean up test files

- Remove commented-out code and unused TODOs
- Improve code formatting and style consistency

aissat committed:
Update SDK constraints and add matcher dependency

aissat committed:
Fix empty string handling in capitalize and remove unused dependency
